From 16d2d495cd10b717f8bf2689d533e0ca84f2ddc5 Mon Sep 17 00:00:00 2001
From: Rahul Singh <rahul.singh@arm.com>
Date: Thu, 16 Jun 2022 15:38:41 +0100
Subject: [PATCH 36/78] xen/arm: pci: Add support for
 pci_generic_config_{read,write}32

pci_generic_config_{read,write}32 support is required for some
host-bridges to access the config space.

Signed-off-by: Rahul Singh <rahul.singh@arm.com>
---
 xen/arch/arm/pci/pci-access.c | 42 +++++++++++++++++++++++++++++++++++
 xen/include/asm-arm/pci.h     |  4 ++++
 2 files changed, 46 insertions(+)

diff --git a/xen/arch/arm/pci/pci-access.c b/xen/arch/arm/pci/pci-access.c
index 9f9aac43d7..cc0f23950a 100644
--- a/xen/arch/arm/pci/pci-access.c
+++ b/xen/arch/arm/pci/pci-access.c
@@ -73,6 +73,48 @@ int pci_generic_config_write(struct pci_host_bridge *bridge, pci_sbdf_t sbdf,
     return 0;
 }
 
+int pci_generic_config_read32(struct pci_host_bridge *bridge, pci_sbdf_t sbdf,
+                              uint32_t where, uint32_t size, uint32_t *val)
+{
+    void __iomem *addr;
+
+    addr = bridge->ops->map_bus(bridge, sbdf, where & ~0x3);
+    if ( !addr ) {
+        *val = ~0;
+        return -ENODEV;
+    }
+
+    *val = readl(addr);
+
+    if ( size <= 2 )
+        *val = (*val >> (8 * (where & 3))) & ((1 << (size * 8)) - 1);
+
+    return 0;
+}
+
+int pci_generic_config_write32(struct pci_host_bridge *bridge, pci_sbdf_t sbdf,
+                               uint32_t where, uint32_t size, uint32_t val)
+{
+    void __iomem *addr;
+    u32 mask, tmp;
+
+    addr = bridge->ops->map_bus(bridge, sbdf, where & ~0x3);
+    if ( !addr )
+        return -ENODEV;
+
+    if ( size == 4 ) {
+        writel(val, addr);
+        return 0;
+    }
+
+    mask = ~(((1 << (size * 8)) - 1) << ((where & 0x3) * 8));
+    tmp = readl(addr) & mask;
+    tmp |= val << ((where & 0x3) * 8);
+    writel(tmp, addr);
+
+    return 0;
+}
+
 static uint32_t pci_config_read(pci_sbdf_t sbdf, unsigned int reg,
                                 unsigned int len)
 {
diff --git a/xen/include/asm-arm/pci.h b/xen/include/asm-arm/pci.h
index 0fc6b6c56a..c627f390d6 100644
--- a/xen/include/asm-arm/pci.h
+++ b/xen/include/asm-arm/pci.h
@@ -101,6 +101,10 @@ int pci_generic_config_read(struct pci_host_bridge *bridge, pci_sbdf_t sbdf,
                             uint32_t reg, uint32_t len, uint32_t *value);
 int pci_generic_config_write(struct pci_host_bridge *bridge, pci_sbdf_t sbdf,
                              uint32_t reg, uint32_t len, uint32_t value);
+int pci_generic_config_read32(struct pci_host_bridge *bridge, pci_sbdf_t sbdf,
+                              uint32_t where, uint32_t size, uint32_t *val);
+int pci_generic_config_write32(struct pci_host_bridge *bridge, pci_sbdf_t sbdf,
+                               uint32_t where, uint32_t size, uint32_t val);
 void __iomem *pci_ecam_map_bus(struct pci_host_bridge *bridge,
                                pci_sbdf_t sbdf, uint32_t where);
 bool pci_ecam_need_p2m_hwdom_mapping(struct domain *d,
-- 
2.25.1

